src/linear_algebra/symplectic_group.lean

```diff
@@ -0,0 +1,252 @@
+/-
+Copyright (c) 2022 Matej Penciak. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Matej Penciak, Moritz Doll, Fabien Clery
+-/
+
+import data.real.basic
+import linear_algebra.matrix.nonsingular_inverse
+
+/-!
+# The Symplectic Group
+
+This file defines the symplectic group and proves elementary properties.
+
+## Main Definitions
+
+`matrix.J`: the canonical `2n × 2n` skew-symmetric matrix
+`symplectic_group`: the group of symplectic matrices
+
+## Implementation Notes
+
+## TODO
+* Every symplectic matrix has determinant 1.
+* For `n = 1` the symplectic group coincides with the special linear group.
+-/
+
+open_locale matrix
+
+variables {l R : Type*}
+
+lemma pm_one_unit [ring R] {x : R} (h : x = 1 ∨ x = -1) : is_unit x :=
+begin
+  cases h,
+  { simp only [h, is_unit_one] },
+  { use -1,
+    simp only [h, units.coe_neg_one] }
+end
+
+section J_matrix_lemmas
+
+namespace matrix
+
+variables (l) [decidable_eq l]
+
+variables (R) [comm_ring R]
+
+/-- The matrix defining the canonical skew-symmetric bilinear form. -/
+def J : matrix (l ⊕ l) (l ⊕ l) R := matrix.from_blocks 0 (-1) 1 0
+
+@[simp] lemma J_transpose : (J l R)ᵀ = - (J l R) :=
+begin
+  unfold J,
+  rw [from_blocks_transpose, ←neg_one_smul R (from_blocks _ _ _ _), from_blocks_smul,
+    matrix.transpose_zero, matrix.transpose_one, transpose_neg],
+  simp [from_blocks],
+end
+
+variables [fintype l]
+
+lemma J_squared : (J l R) ⬝ (J l R) = -1 :=
+begin
+  unfold J,
+  rw from_blocks_multiply,
+  simp only [matrix.zero_mul, matrix.neg_mul, zero_add, neg_zero', matrix.one_mul, add_zero],
+  rw [← neg_zero, ← matrix.from_blocks_neg,← from_blocks_one],
+end
+
+lemma J_inv : (J l R)⁻¹ = -(J l R) :=
+begin
+  refine matrix.inv_eq_right_inv _,
+  rw [matrix.mul_neg, J_squared],
+  exact neg_neg 1,
+end
+
+variables (R)
+variables [no_zero_divisors R] [nontrivial R]
+
+lemma J_det [rc : fact (ring_char R ≠ 2)] : det (J l R) = 1 ∨ det (J l R) = - 1:=
+begin
+  have H : (det (J l R)) * (det (J l R)) = 1 := by
+  { rw [←det_mul, J_squared],
+    rw [←one_smul R (-1 : matrix _ _ R)],
+    rw [smul_neg, ←neg_smul, det_smul],
+    simp only [fintype.card_sum, det_one, mul_one],
+    rw neg_one_pow_eq_one_iff_even (ring.neg_one_ne_one_of_char_ne_two (fact.elim rc)),
+    exact even_add_self _ },
+  rw [←sq_eq_one_iff, pow_two],
+  exact H,
+end
+
+lemma J_det_unit [fact (ring_char R ≠ 2)] : is_unit (det (J l R)) :=
+  pm_one_unit (J_det l R)
+
+end matrix
+
+end J_matrix_lemmas
+
+open matrix
+
+variables (l) (R) [decidable_eq l] [fintype l] [comm_ring R]
+
+/-- The group of symplectic matrices over a ring `R`. -/
+def symplectic_group : submonoid (matrix (l ⊕ l) (l ⊕ l)  R) :=
```

Comment:
This should maybe be called `matrix.symplectic_group` (with the leading namespace), since presumably there are representations with linear maps too?

Reply:
Yes, there is an abstract version with a non-degenerate skew-symmetric bilinear form and a version of Gram-Schmidt that connects the two versions.